analyzers/src/SonarAnalyzer.CSharp/Helpers/CSharpSymbolUsageCollector.cs

```diff
@@ -103,26 +86,28 @@ public override void VisitAssignmentExpression(AssignmentExpressionSyntax node)
 
             base.VisitAssignmentExpression(node);
 
-            static INamedTypeSymbol GetNamedTypeSymbol(ISymbol symbol)
+            static int GetTupleCount(ExpressionSyntax assignmentLeft)
             {
-                if (symbol != null)
+                var result = 0;
+                if (TupleExpressionSyntaxWrapper.IsInstance(assignmentLeft))
                 {
-                    if (symbol.GetSymbolType() is INamedTypeSymbol namedType)
-                    {
-                        return namedType;
-                    }
-                    else if (symbol is IMethodSymbol methodSymbol)
-                    {
-                        return methodSymbol.MethodKind == MethodKind.Constructor
-                            ? methodSymbol.ContainingType
-                            : methodSymbol.ReturnType as INamedTypeSymbol;
-                    }
+                    result = ((TupleExpressionSyntaxWrapper)assignmentLeft).Arguments.Count;
                 }
-                return null;
+                else if (DeclarationExpressionSyntaxWrapper.IsInstance(assignmentLeft)
+                    && (DeclarationExpressionSyntaxWrapper)assignmentLeft is { } leftDeclaration
+                    && ParenthesizedVariableDesignationSyntaxWrapper.IsInstance(((DeclarationExpressionSyntaxWrapper)assignmentLeft).Designation))
+                {
+                    var declarationExpression = (ParenthesizedVariableDesignationSyntaxWrapper)((DeclarationExpressionSyntaxWrapper)assignmentLeft).Designation;
+                    result = declarationExpression.Variables.Count;
+                }
+                return result;
             }
 
-            ISymbol FindDeconstructor(IEnumerable<ISymbol> deconstructors, int numberOfArguments) =>
-                deconstructors.FirstOrDefault(m => m.GetParameters().Count() == numberOfArguments);
+            static ISymbol FindDeconstructor(IEnumerable<ISymbol> deconstructors, int numberOfArguments) =>
+                deconstructors.FirstOrDefault(m => m.GetParameters().Count() == numberOfArguments && IsAccessibleOutsideTheType(m.DeclaredAccessibility));
+
+            static bool IsAccessibleOutsideTheType(Accessibility accessibility) =>
```

Comment:
@pavel-mikula-sonarsource  I tried to refactor in `AccessibilityExtensions` this one and the one you added in #3837 in `MemberShouldBeStatic.cs`, but there's a difference:
- here, I want `ProtectedAndInternal`
- there, we use `ProtectedOrInternal`

After the refactoring, a test of `MemberShouldBeStatic` broke so I just stopped.

Reply:
I will take a look later.

Reply:
I checked the S2325 and `protected internal Method()` has `DeclaredAccessibility` with `Accessibility.ProtectedOrInternal`. So S2325 implementation is fine.

Reply:
Hmm, you have `internal` class with `protected internal` member. I have `public` class with `protected internal` member.

So we are both right and both wrong at the same time :D

Reply:
Final version: `ProtectedAndInternal` is for `private protected` C# 7 construct. Fix is coming in #3862